include/elements/3d/tetrahedron_element.tcc

```diff
@@ -0,0 +1,396 @@
+//! to check if a point is inside /outside of a hedron
+//! \retval indices Indices that form sub-tetrahedrons
+template <unsigned Tdim, unsigned Tnfunctions>
+inline Eigen::MatrixXi
+    mpm::TetrahedronElement<Tdim, Tnfunctions>::inhedron_indices() const {
+  Eigen::Matrix<int, 6, Tdim, Eigen::RowMajor> indices;
+  // clang-format off
+  indices << 0, 1,
+             1, 2,
+             2, 3,
+             3, 0,
+             1, 3,
+             0, 2;
+  //clang-format on
+  return indices;
+}
+
+//! Return indices of a face of the element
+//! 4-noded tetrahedron
+template <>
+inline Eigen::VectorXi
+    mpm::TetrahedronElement<3, 4>::face_indices(unsigned face_id) const {
+
+  //! Face ids and its associated nodal indices
+  const std::map<unsigned, Eigen::Matrix<int, 3, 1>> face_indices_tetrahedron{
+      {0, Eigen::Matrix<int, 3, 1>(0, 1, 2)},
+      {1, Eigen::Matrix<int, 3, 1>(0, 1, 3)},
+      {2, Eigen::Matrix<int, 3, 1>(0, 2, 3)},
+      {3, Eigen::Matrix<int, 3, 1>(1, 2, 3)}};
+
+  return face_indices_tetrahedron.at(face_id);
+}
+
+//! Return quadrature
+template <unsigned Tdim, unsigned Tnfunctions>
+inline std::shared_ptr<mpm::Quadrature<Tdim>>
+    mpm::TetrahedronElement<Tdim, Tnfunctions>::quadrature(
+        unsigned nquadratures) const {
+  switch (nquadratures) {
+    case 1:
+      return Factory<mpm::Quadrature<Tdim>>::instance()->create("QTET1");
+      break;
+    case 2:
+      return Factory<mpm::Quadrature<Tdim>>::instance()->create("QTET2");
+      break;
+    default:
+      return Factory<mpm::Quadrature<Tdim>>::instance()->create("QTET1");
+      break;
+  }
+}
+
+//! Compute volume
+//! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+//! \retval volume Return the volume of cell
+template <unsigned Tdim, unsigned Tnfunctions>
+inline double mpm::TetrahedronElement<Tdim, Tnfunctions>::compute_volume(
+    const Eigen::MatrixXd& nodal_coordinates) const {
+
+  //Temporary variables for node coordinants (xk, yk, zk)
+  const double x1 = nodal_coordinates(0, 0);
+  const double x2 = nodal_coordinates(1, 0);
+  const double x3 = nodal_coordinates(2, 0);
+  const double x4 = nodal_coordinates(3, 0);
+  const double y1 = nodal_coordinates(0, 1);
+  const double y2 = nodal_coordinates(1, 1);
+  const double y3 = nodal_coordinates(2, 1);
+  const double y4 = nodal_coordinates(3, 1);
+  const double z1 = nodal_coordinates(0, 2);
+  const double z2 = nodal_coordinates(1, 2);
+  const double z3 = nodal_coordinates(2, 2);
+  const double z4 = nodal_coordinates(3, 2);
+
+  const double volume =
+    (1.0 / 6)*
+    ( x1 * y3 * z2 - x1 * y2 * z3 + x2 * y1 * z3 - x2 * y3 * z1 -
+    x3 * y1 * z2 + x3 * y2 * z1 + x1 * y2 * z4 - x1 * y4 * z2 -
+    x2 * y1 * z4 + x2 * y4 * z1 + x4 * y1 * z2 - x4 * y2 * z1 -
+    x1 * y3 * z4 + x1 * y4 * z3 + x3 * y1 * z4 - x3 * y4 * z1 -
+    x4 * y1 * z3 + x4 * y3 * z1 + x2 * y3 * z4 - x2 * y4 * z3 -
+    x3 * y2 * z4 + x3 * y4 * z2 + x4 * y2 * z3 - x4 * y3 * z2 );
```

Comment:
I think I missed some converstion about this section... what was the idea with not using the dot and cross product of the vector repsentation of the nodal coordinates? (e.g., `\vec{x}_0 = nodal_coordinates(:, 0)`

Reply:
I'm pretty sure there is a form using only dot and cross product (no det) for volume. I don't think the vector multiplication would be any more expensive than whats above

Reply:
After thinking about it again, I think the cost of storing x1 x2 x3 x4 ... etc is ok. Actually, it is going to cost us local storage, but since we make them `const`, I think the access and storage costs are not so expensive. I am thinking that it might be more expensive to access an element in the array `nodal_coordinates()` over time.

Reply:
For the second comment, I am ok with anyway: direct multiplication (like Lauren did) or even using `det` function in Eigen. I believe they have a fast calculation module for determinant.